sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java

```diff
@@ -0,0 +1,177 @@
+      PCollectionTuple pct;
+
+      /**
+       * Create DicomIO.ReadDicomStudyMetadata.Result from PCollectionTuple which contains the
+       * response (with METADATA and ERROR_MESSAGE tags).
+       *
+       * @param pct the pct
+       * @return the read result
+       * @throws IllegalArgumentException the illegal argument exception
+       */
+      static DicomIO.ReadDicomStudyMetadata.Result of(PCollectionTuple pct)
+          throws IllegalArgumentException {
+        if (pct.getAll()
+            .keySet()
+            .containsAll((Collection<?>) TupleTagList.of(METADATA).and(ERROR_MESSAGE))) {
+          return new DicomIO.ReadDicomStudyMetadata.Result(pct);
+        } else {
+          throw new IllegalArgumentException(
+              "The PCollection tuple must have the DicomIO.ReadDicomStudyMetadata.METADATA "
+                  + "and DicomIO.ReadDicomStudyMetadata.ERROR_MESSAGE tuple tags");
+        }
+      }
+
+      private Result(PCollectionTuple pct) {
+        this.pct = pct;
+        this.readResponse = pct.get(METADATA);
+        this.failedReads = pct.get(ERROR_MESSAGE);
+      }
+
+      /**
+       * Gets failed reads.
+       *
+       * @return the failed reads
+       */
+      public PCollection<String> getFailedReads() {
+        return failedReads;
+      }
+
+      /**
+       * Gets resources.
+       *
+       * @return the resources
+       */
+      public PCollection<String> getReadResponse() {
+        return readResponse;
+      }
+
+      @Override
+      public Pipeline getPipeline() {
+        return this.pct.getPipeline();
+      }
+
+      @Override
+      public Map<TupleTag<?>, PValue> expand() {
+        return ImmutableMap.of(METADATA, readResponse);
+      }
+
+      @Override
+      public void finishSpecifyingOutput(
+          String transformName, PInput input, PTransform<?, ?> transform) {}
+    }
+
+    /**
+     * DoFn to fetch the metadata of a study from a Dicom store based on it's location and study id.
+     */
+    static class FetchStudyMetadataFn extends DoFn<PubsubMessage, String> {
+
+      private HealthcareApiClient dicomStore;
+
+      /**
+       * Instantiate the healthcare client.
+       *
+       * @throws IOException
+       */
+      @Setup
+      public void instantiateHealthcareClient() throws IOException {
+        this.dicomStore = new HttpHealthcareApiClient();
+      }
+
+      /**
+       * Process The Pub/Sub message.
+       *
+       * @param context The input containing the pub/sub message
+       */
+      @ProcessElement
+      public void processElement(ProcessContext context) {
+        PubsubMessage msg = context.element();
+        byte[] msgPayload = msg.getPayload();
+        try {
+          String dicomWebPath = new String(msgPayload, "UTF-8");
+          String responseData = dicomStore.retrieveDicomStudyMetadata(dicomWebPath);
+          context.output(METADATA, responseData);
```

Comment:
I see that you're making individual blocking requests to the dicomStore - I have a couple questions:
- Does dicom support batched requests?
- Does it support concurrent requests?

In this case, you're issuing individual, blocking calls, which may give your transform a very low throughput. If you can use batching and/or async IO, you should be able to reach a higher throughput.

Reply:
The Dicom API does support async calls. As you suggested, it is probably better to make use of this. 
However, the read metadata endpoint is generally very fast, as it is transferring little data. So in the interest of time, would it be okay to leave it for now, and return to this in a later feature request?

Reply:
okay can you add a TODO there?

Reply:
Sure thing